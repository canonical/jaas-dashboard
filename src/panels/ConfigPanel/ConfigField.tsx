import { ReactNode, useEffect, useRef, useState } from "react";
import classnames from "classnames";

import { isSet } from "components/utils";
import { ConfigData } from "juju/api";

export type SetNewValue = (name: string, value: any) => void;

export type ConfigProps = {
  config: ConfigData;
  selectedConfig: ConfigData | undefined;
  setSelectedConfig: Function;
  setNewValue: SetNewValue;
};

type Props<V> = ConfigProps & {
  input: (value: V) => ReactNode;
};

const ConfigField = <V,>({
  config,
  input,
  selectedConfig,
  setSelectedConfig,
  setNewValue,
}: Props<V>): JSX.Element => {
  const [inputFocused, setInputFocused] = useState(false);
  const [inputChanged, setInputChanged] = useState(false);
  const [showUseDefault, setShowUseDefault] = useState(
    config.value !== config.default
  );
  const [showDescription, setShowDescription] = useState(false);
  const descriptionRef = useRef<HTMLDivElement>(null);
  const [maxDescriptionHeight, setMaxDescriptionHeight] = useState("0px");

  let inputValue = config.default;
  if (isSet(config.newValue)) {
    inputValue = config.newValue;
  } else if (config.default !== config.value) {
    inputValue = config.value;
  }

  useEffect(() => {
    if (descriptionRef.current?.firstChild) {
      setMaxDescriptionHeight(
        `${
          (descriptionRef.current.firstChild as HTMLPreElement).clientHeight
        }px`
      );
    }
  }, []);

  useEffect(() => {
    if (!descriptionRef.current) {
      return;
    }
    if (showDescription) {
      descriptionRef.current.style.maxHeight = maxDescriptionHeight;
    } else {
      descriptionRef.current.style.maxHeight = "0px";
    }
  }, [showDescription, maxDescriptionHeight]);

  useEffect(() => {
    setInputFocused(selectedConfig?.name === config.name);
  }, [selectedConfig, config]);

  useEffect(() => {
    if (
      (isSet(config.newValue) && config.newValue !== config.default) ||
      (!isSet(config.newValue) && config.value !== config.default)
    ) {
      setShowUseDefault(true);
    } else {
      setShowUseDefault(false);
    }

    if (isSet(config.newValue) && config.newValue !== config.value) {
      setInputChanged(true);
    } else {
      setInputChanged(false);
    }
  }, [config]);

  function resetToDefault() {
    setNewValue(config.name, config.default);
  }

  function handleShowDescription() {
    setShowDescription(!showDescription);
  }

  return (
    // XXX How to tell aria to ignore the click but not the element?
    // eslint-disable-next-line
    <div
      className={classnames("config-input", {
        "config-input--focused": inputFocused,
        "config-input--changed": inputChanged,
      })}
      data-testid={config.name}
      onClick={() => setSelectedConfig(config)}
    >
      <h5 className="u-float-left">
        <i
          className={classnames("config-input--view-description", {
            "p-icon--plus": !showDescription,
            "p-icon--minus": showDescription,
          })}
          onClick={handleShowDescription}
          onKeyPress={handleShowDescription}
          role="button"
          tabIndex={0}
        />
        {config.name}
      </h5>
      <button
        className={classnames(
          "u-float-right p-button--base config-panel__hide-button",
          {
            "config-panel__show-button": showUseDefault,
          }
        )}
        onClick={resetToDefault}
      >
        use default
      </button>
      <div
        className={classnames("config-input--description")}
        ref={descriptionRef}
      >
        <pre className="config-input--description-container">
          {config.description}
        </pre>
      </div>
      {input(inputValue)}
    </div>
  );
};

export default ConfigField;
