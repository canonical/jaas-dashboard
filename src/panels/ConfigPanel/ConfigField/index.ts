export { default } from "./ConfigField";
export { Label as ConfigFieldLabel } from "./types";
export type {
  ConfigProps,
  SetNewValue,
  SetSelectedConfig,
} from "./ConfigField";
