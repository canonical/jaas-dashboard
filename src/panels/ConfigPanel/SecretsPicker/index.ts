export { default } from "./SecretsPicker";
export { Label as SecretsPickerLabel } from "./types";
