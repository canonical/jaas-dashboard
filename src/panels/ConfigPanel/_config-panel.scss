@import "vanilla-framework/scss/vanilla";
@import "../../scss/settings";
@import "../../scss/breakpoints";
@import "../../scss/variables";

$panel-padding: 1.5rem;

.slide-panel.config-panel .slide-panel__content {
  padding: $panel-padding 0 0;
}

.config-panel {
  display: flex;
  flex: 1;
  flex-direction: column;

  &__modal-button-row-hint {
    font-size: #{map-get($font-sizes, small)}rem;
    margin-bottom: 1.5rem;
    text-align: left;
  }

  h5 {
    margin-bottom: 0.2rem;
  }

  pre {
    // reset the vanilla pre styling
    background-color: inherit;
    border: none;
    padding: 0;
    // extra styles
    white-space: pre-wrap;
    word-break: keep-all;
  }

  .full-size {
    display: flex;
    justify-content: center;
    text-align: center;
  }

  .config-panel__hide-button {
    @include vf-animation(opacity, brisk, out);

    opacity: 0;
    white-space: nowrap;
    width: auto;

    &[aria-hidden="true"] {
      width: 0;
    }
  }

  .config-panel__show-button {
    opacity: 1;
  }

  &__reset-all {
    flex: 1 0 auto;
    margin-left: auto;
    text-align: right;

    &.config-panel__hide-button {
      overflow: hidden;
      position: absolute;
      right: $sph--large * 2;
      top: 0;
    }
  }

  &__message {
    padding-top: 6rem;

    @include desktop {
      padding-top: 3rem;
    }

    h4 {
      margin-top: 3rem;
    }
  }

  &--center-img {
    margin: 0 auto;
  }

  &__list {
    padding-right: $sph--large * 2;
  }

  .p-panel__content {
    flex: 1;

    .aside-split-col {
      display: flex;
      flex-direction: column;

      .config-panel__list {
        flex: 1;
      }
    }
  }

  &__list-header {
    display: flex;
    justify-content: space-between;
    margin-bottom: $spv--large;
    padding: 0 $sph--large * 2 0 $sph--small;
    position: relative;

    h5 {
      // Need to leave extra room for the reset button so that when it appears/disappears
      // it does not move the content around.
      padding-bottom: $sph--small;
    }

    @media (max-width: $breakpoint-x-small) {
      flex-direction: column;
    }

    @include mobile {
      margin-top: 4rem;
      padding-right: 4rem; // Spacing for the close X
      padding-top: 1rem;
    }

    button {
      margin-bottom: 0;
      margin-left: 1rem;
      padding-left: 0.5rem;
      padding-right: 0.5rem;
      white-space: nowrap;

      @media screen and (max-width: $breakpoint-x-small) {
        margin-left: inherit;
      }
    }
  }

  &__description {
    @media screen and (max-width: $breakpoint-small) {
      display: none !important;
    }

    &-wrapper {
      position: sticky;
      top: 1rem;
    }
  }

  &__no-description {
    padding: 0 5rem;
    position: sticky;
    text-align: center;
    top: 0;
  }

  &__drawer {
    background-color: white;
    bottom: 0;
    box-shadow: none;
    position: sticky;

    @media screen and (max-width: $breakpoint-x-small) {
      padding-right: 0;
    }

    &.is-open {
      box-shadow: $box-shadow;
    }
  }

  &__button-row {
    border-top: 1px solid $color-mid-light;
    margin-right: $sph--large;
    text-align: right;

    button {
      margin-top: 1.2rem;
    }

    button.config-panel__save-button {
      @include vf-animation(opacity, brisk, out);

      span {
        padding-left: 1rem;
      }
    }
  }
}

.config-input {
  border: 1px solid transparent;
  border-radius: 2px;
  border-radius: $border-radius;
  margin-bottom: $spv--small;
  padding: 0 0.5rem;

  &--focused {
    @include vf-animation(#{background-color, border-color}, brisk, out);

    background-color: $color-blue-highlight-background;
    border-color: $color-blue-highlight-border;
  }

  &--changed:not(.config-input--focused) {
    background-color: $color-light;
  }

  &__title {
    display: flex;
  }

  .config-input--view-description {
    display: none;
    // Give the toggle button a larger hit area:
    margin: ($spv--x-small * 0.5) 0 0 (-$spv--small);
    padding: $spv--x-small $spv--small;

    .p-icon--minus,
    .p-icon--plus {
      width: #{map-get($font-sizes, small)}rem;
    }

    @media screen and (max-width: $breakpoint-small) {
      display: inline-block;
    }
  }

  &--description {
    @include vf-animation(max-height, brisk, out);

    clear: both;
    display: none;
    font-size: #{map-get($font-sizes, small)}rem;
    margin-bottom: 0.5rem;
    margin-left: 1.3rem;
    max-height: 0;
    overflow: hidden;

    @media screen and (max-width: $breakpoint-large) {
      display: block;
    }
  }

  button {
    color: $color-link;
    margin: 0.3rem 0 0;
    padding: 0;

    &:hover {
      background-color: inherit;
    }
  }

  @media screen and (max-width: $breakpoint-x-small) {
    button {
      width: auto;
    }
  }

  @media screen and (min-width: $breakpoint-large) {
    textarea {
      margin-bottom: 1.2rem !important;
    }
  }

  textarea {
    border: 1px solid $color-grey60;
    border-radius: 2px;
    height: 2.3rem;
    margin-bottom: 0.7rem;
    min-height: 2.3rem;
    resize: vertical;
  }
}
