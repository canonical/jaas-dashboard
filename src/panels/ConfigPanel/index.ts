export { default } from "./ConfigPanel";
export {
  Label as ConfigPanelLabel,
  TestId as ConfigPanelTestId,
} from "./types";
