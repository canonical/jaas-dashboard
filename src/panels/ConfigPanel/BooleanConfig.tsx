import { RadioInput } from "@canonical/react-components";
import ConfigField, { ConfigProps } from "./ConfigField";

export default function BooleanConfig({
  config,
  selectedConfig,
  setSelectedConfig,
  setNewValue,
}: ConfigProps): JSX.Element {
  function handleOptionChange(e: any) {
    const bool = e.target.value === "true" ? true : false;
    setNewValue(e.target.name, bool);
  }

  return (
    <ConfigField<boolean>
      config={config}
      selectedConfig={selectedConfig}
      setSelectedConfig={setSelectedConfig}
      setNewValue={setNewValue}
      input={(value) => (
        <div className="row">
          <div className="col-2">
            <RadioInput
              label="true"
              name={config.name}
              aria-labelledby={config.name}
              checked={value === true}
              value="true"
              onClick={handleOptionChange}
              onChange={handleOptionChange}
            />
          </div>
          <div className="col-2">
            <RadioInput
              label="false"
              name={config.name}
              aria-labelledby={config.name}
              checked={value === false}
              value="false"
              onClick={handleOptionChange}
              onChange={handleOptionChange}
            />
          </div>
        </div>
      )}
    />
  );
}
