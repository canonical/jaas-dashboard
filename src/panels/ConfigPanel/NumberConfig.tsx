import ConfigField, { ConfigProps } from "./ConfigField";

const NumberConfig = ({
  config,
  selectedConfig,
  setSelectedConfig,
  setNewValue,
}: ConfigProps): JSX.Element => {
  return (
    <ConfigField<number>
      config={config}
      selectedConfig={selectedConfig}
      setSelectedConfig={setSelectedConfig}
      setNewValue={setNewValue}
      input={(value) => (
        <input
          type="number"
          value={value}
          onFocus={() => setSelectedConfig(config)}
          onChange={(e) => {
            setNewValue(config.name, e.target.value);
          }}
        />
      )}
    />
  );
};

export default NumberConfig;
