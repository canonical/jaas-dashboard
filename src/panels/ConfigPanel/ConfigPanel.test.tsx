import { screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import * as apiModule from "juju/api";
import type { RootState } from "store/store";
import {
  applicationGetFactory,
  configFactory,
} from "testing/factories/juju/Application";
import { modelUserInfoFactory } from "testing/factories/juju/ModelManagerV9";
import {
  controllerFactory,
  jujuStateFactory,
  modelDataFactory,
  modelDataInfoFactory,
} from "testing/factories/juju/juju";
import { rootStateFactory } from "testing/factories/root";
import { renderComponent } from "testing/utils";

import ConfigPanel, { Label } from "./ConfigPanel";

jest.mock("juju/api", () => ({
  getApplicationConfig: jest.fn(),
  setApplicationConfig: jest.fn(),
}));

describe("ConfigPanel", () => {
  let state: RootState;
  const params = new URLSearchParams({
    entity: "easyrsa",
    charm: "cs:easyrsa",
    modelUUID: "abc123",
    panel: "config",
  });
  const url = `/models/eggman@external/hadoopspark?${params.toString()}`;
  const path = "/models/:userName/:modelName";
  let getApplicationConfigSpy: jest.SpyInstance;
  const consoleError = console.error;

  beforeEach(() => {
    console.error = jest.fn();
    jest.resetModules();
    state = rootStateFactory.build({
      juju: jujuStateFactory.build({
        controllers: {
          "wss://jimm.jujucharms.com/api": [
            controllerFactory.build({ path: "admin/jaas", uuid: "123" }),
          ],
        },
        modelData: {
          abc123: modelDataFactory.build({
            info: modelDataInfoFactory.build({
              "controller-uuid": "123",
              name: "hadoopspark",
              users: [
                modelUserInfoFactory.build({ user: "eggman@external" }),
                modelUserInfoFactory.build({ user: "spaceman@domain" }),
              ],
            }),
          }),
        },
      }),
    });
    getApplicationConfigSpy = jest
      .spyOn(apiModule, "getApplicationConfig")
      .mockImplementation(() =>
        Promise.resolve(
          applicationGetFactory.build({
            config: {
              email: configFactory.build({ default: "" }),
              name: configFactory.build({ default: "eggman" }),
            },
          }),
        ),
      );
  });

  afterEach(() => {
    console.error = consoleError;
    jest.restoreAllMocks();
  });

  it("displays a message if the app has no config", async () => {
    jest.spyOn(apiModule, "getApplicationConfig").mockImplementation(() => {
      return Promise.resolve(applicationGetFactory.build({ config: {} }));
    });
    renderComponent(<ConfigPanel />, { state, path, url });
    // Use findBy to wait for the async events to finish
    await screen.findByText(Label.NONE);
    expect(document.querySelector(".config-panel__message")).toMatchSnapshot();
  });

  it("can display boolean, number and text fields", async () => {
    getApplicationConfigSpy = jest
      .spyOn(apiModule, "getApplicationConfig")
      .mockImplementation(() =>
        Promise.resolve(
          applicationGetFactory.build({
            config: {
              name: configFactory.build({ type: "string" }),
              age: configFactory.build({ type: "int" }),
              confirm: configFactory.build({ type: "boolean" }),
            },
          }),
        ),
      );
    renderComponent(<ConfigPanel />, { state, path, url });
    expect(
      within(await screen.findByTestId("name")).getByRole("textbox"),
    ).toBeInTheDocument();
    expect(
      within(await screen.findByTestId("age")).getByRole("spinbutton"),
    ).toBeInTheDocument();
    expect(
      within(await screen.findByTestId("confirm")).getAllByRole("radio"),
    ).toHaveLength(2);
  });

  it("highlights changed fields before save", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    const wrapper = await screen.findByTestId("email");
    const input = within(wrapper).getByRole("textbox");
    await userEvent.type(input, "new value");
    expect(input).toHaveTextContent("new value");
    expect(wrapper).toHaveClass("config-input--changed");
  });

  it("can reset all fields", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    const email = within(await screen.findByTestId("email")).getByRole(
      "textbox",
    );
    const name = within(await screen.findByTestId("name")).getByRole("textbox");
    await userEvent.type(email, "eggman@example.com");
    await userEvent.type(name, "not eggman");
    expect(email).toHaveTextContent("eggman@example.com");
    expect(name).toHaveTextContent("not eggman");
    await userEvent.click(
      screen.getByRole("button", { name: Label.RESET_BUTTON }),
    );
    expect(email).toHaveTextContent("");
    expect(name).toHaveTextContent("eggman");
  });

  it("displays a confirmation when clicking outside and there are unsaved changes", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.click(document.body);
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.CANCEL_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).toBeInTheDocument();
    expect(window.location.search).toBe(`?${params.toString()}`);
  });

  it("closes when clicking outside and there are no unsaved changes", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.click(document.body);
    expect(
      within(screen.getByRole("dialog", { name: "" })).queryByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).not.toBeInTheDocument();
    expect(window.location.search).toBeFalsy();
  });

  it("displays a confirmation when cancelling and there are unsaved changes", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.CANCEL_BUTTON }),
    );
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.CANCEL_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).toBeInTheDocument();
    expect(window.location.search).toBe(`?${params.toString()}`);
  });

  it("can confirm the cancel confirmation", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.CANCEL_BUTTON }),
    );
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.CANCEL_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).toBeInTheDocument();
    expect(window.location.search).toBe(`?${params.toString()}`);
    await userEvent.click(
      screen.getByRole("button", { name: Label.CANCEL_CONFIRM_CONFIRM_BUTTON }),
    );
    expect(window.location.search).toBeFalsy();
  });

  it("can cancel the cancel confirmation", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.CANCEL_BUTTON }),
    );
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.CANCEL_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).toBeInTheDocument();
    expect(window.location.search).toBe(`?${params.toString()}`);
    await userEvent.click(
      screen.getByRole("button", { name: Label.CANCEL_CONFIRM_CANCEL_BUTTON }),
    );
    expect(window.location.search).toBe(`?${params.toString()}`);
  });

  it("closes when cancelling and there are no unsaved changes", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.click(
      await screen.findByRole("button", { name: Label.CANCEL_BUTTON }),
    );
    expect(
      within(screen.getByRole("dialog", { name: "" })).queryByRole("heading", {
        name: Label.CANCEL_CONFIRM,
      }),
    ).not.toBeInTheDocument();
    expect(window.location.search).toBeFalsy();
  });

  it("displays a confirmation before saving", async () => {
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.type(
      within(await screen.findByTestId("name")).getByRole("textbox"),
      "noteggman",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_BUTTON }),
    );
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.SAVE_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.SAVE_CONFIRM,
      }),
    ).toBeInTheDocument();
  });

  it("can cancel the save confirmation", async () => {
    const setApplicationConfigSpy = jest.spyOn(
      apiModule,
      "setApplicationConfig",
    );
    renderComponent(<ConfigPanel />, { state, path, url });
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.type(
      within(await screen.findByTestId("name")).getByRole("textbox"),
      "noteggman",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_BUTTON }),
    );
    expect(
      within(
        screen.getByRole("dialog", {
          name: Label.SAVE_CONFIRM,
        }),
      ).getByRole("heading", {
        name: Label.SAVE_CONFIRM,
      }),
    ).toBeInTheDocument();
    await userEvent.click(
      within(
        screen.getByRole("dialog", {
          name: Label.SAVE_CONFIRM,
        }),
      ).getByRole("button", {
        name: Label.SAVE_CONFIRM_CANCEL_BUTTON,
      }),
    );
    expect(screen.queryByRole("dialog", { name: "" })).not.toBeInTheDocument();
    expect(setApplicationConfigSpy).not.toHaveBeenCalled();
  });

  it("can save changes", async () => {
    const setApplicationConfigSpy = jest
      .spyOn(apiModule, "setApplicationConfig")
      .mockImplementation(() => Promise.resolve({ results: [] }));
    renderComponent(<ConfigPanel />, { state, path, url });
    expect(getApplicationConfigSpy).toHaveBeenCalledTimes(1);
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.type(
      within(await screen.findByTestId("name")).getByRole("textbox"),
      "noteggman",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_BUTTON }),
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_CONFIRM_CONFIRM_BUTTON }),
    );
    expect(setApplicationConfigSpy).toHaveBeenCalledWith(
      "abc123",
      "easyrsa",
      {
        email: configFactory.build({
          name: "email",
          default: "",
          newValue: "eggman@example.com",
        }),
        name: configFactory.build({
          name: "name",
          default: "eggman",
          newValue: "noteggman",
        }),
      },
      state,
    );
    expect(getApplicationConfigSpy).toHaveBeenCalledTimes(2);
  });

  it("displays save errors", async () => {
    jest.spyOn(apiModule, "setApplicationConfig").mockImplementation(() =>
      Promise.resolve({
        results: [{ error: { code: "1", message: "That's not a name" } }],
      }),
    );
    renderComponent(<ConfigPanel />, { state, path, url });
    expect(getApplicationConfigSpy).toHaveBeenCalledTimes(1);
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.type(
      within(await screen.findByTestId("name")).getByRole("textbox"),
      "noteggman",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_BUTTON }),
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_CONFIRM_CONFIRM_BUTTON }),
    );
    expect(screen.getByText("That's not a name")).toBeInTheDocument();
    expect(getApplicationConfigSpy).toHaveBeenCalledTimes(1);
  });

  it("should display console error when trying to get config", async () => {
    jest
      .spyOn(apiModule, "getApplicationConfig")
      .mockImplementation(
        jest
          .fn()
          .mockRejectedValue(
            new Error("Error while calling getApplicationConfig"),
          ),
      );
    renderComponent(<ConfigPanel />, { state, path, url });
    expect(apiModule.getApplicationConfig).toHaveBeenCalledTimes(1);
    await waitFor(() =>
      expect(console.error).toHaveBeenCalledWith(
        Label.GET_CONFIG_ERROR,
        new Error("Error while calling getApplicationConfig"),
      ),
    );
  });

  it("should display console error when trying to save", async () => {
    jest.spyOn(apiModule, "getApplicationConfig").mockImplementationOnce(
      jest.fn().mockResolvedValue(
        applicationGetFactory.build({
          config: {
            email: configFactory.build({ default: "" }),
            name: configFactory.build({ default: "eggman" }),
          },
        }),
      ),
    );
    const setApplicationConfigSpy = jest
      .spyOn(apiModule, "setApplicationConfig")
      .mockImplementation(() =>
        Promise.reject(new Error("Error while trying to save")),
      );
    renderComponent(<ConfigPanel />, { state, path, url });
    expect(getApplicationConfigSpy).toHaveBeenCalledTimes(1);
    await userEvent.type(
      within(await screen.findByTestId("email")).getByRole("textbox"),
      "eggman@example.com",
    );
    await userEvent.type(
      within(await screen.findByTestId("name")).getByRole("textbox"),
      "noteggman",
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_BUTTON }),
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.SAVE_CONFIRM_CONFIRM_BUTTON }),
    );
    expect(setApplicationConfigSpy).toHaveBeenCalledWith(
      "abc123",
      "easyrsa",
      {
        email: configFactory.build({
          name: "email",
          default: "",
          newValue: "eggman@example.com",
        }),
        name: configFactory.build({
          name: "name",
          default: "eggman",
          newValue: "noteggman",
        }),
      },
      state,
    );
    await waitFor(() =>
      expect(console.error).toHaveBeenCalledWith(
        Label.SUBMIT_TO_JUJU_ERROR,
        new Error("Error while trying to save"),
      ),
    );
  });
});
