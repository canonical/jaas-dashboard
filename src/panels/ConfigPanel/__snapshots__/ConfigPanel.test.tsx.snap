// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigPanel displays a message if the app has no config 1`] = `
<div
  class="config-panel__message"
>
  <img
    alt=""
    class="config-panel--center-img"
    src="no-config-params.svg"
  />
  <h4>
    This application doesn't have any configuration parameters
  </h4>
</div>
`;
